enumerables.rb

```diff
@@ -0,0 +1,264 @@
+# rubocop:disable Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+
+module Enumerable
+  # my_each
+  def my_each()
+    return to_enum(:my_each) unless block_given?
+
+    size.times do |i|
+      yield to_a[i]
+    end
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index()
+    return to_enum(:my_each_with_index) unless block_given?
+
+    my_each do |i|
+      yield i, to_a.index(i)
+    end
+    self
+  end
+
```

Comment:
When testing this, the following error occurs:

```
  1) enumerables #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: 4, idx: 0\nNum: 4, idx: 1\nNum: 7, idx: 2\nNum: 2, idx: 3\nNum: 8, idx: 4\nNum: 8, idx: 5\nNum:...: 1, idx: 94\nNum: 4, idx: 95\nNum: 7, idx: 96\nNum: 4, idx: 97\nNum: 7, idx: 98\nNum: 6, idx: 99\n"
            got: "Num: 4, idx: 0\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 2, idx: 3\nNum: 8, idx: 4\nNum: 8, idx: 4\nNum:...\nNum: 1, idx: 16\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 6, idx: 6\n"
     
       (compared using ==)
     
       Diff:
       
       
       
       
       
       @@ -1,101 +1,101 @@
        Num: 4, idx: 0
       -Num: 4, idx: 1
       +Num: 4, idx: 0
        Num: 7, idx: 2
        Num: 2, idx: 3
        Num: 8, idx: 4
       -Num: 8, idx: 5
       +Num: 8, idx: 4
        Num: 6, idx: 6
       -Num: 2, idx: 7
       -Num: 7, idx: 8
       -Num: 7, idx: 9
       -Num: 6, idx: 10
       +Num: 2, idx: 3
       +Num: 7, idx: 2
       +Num: 7, idx: 2
       +Num: 6, idx: 6
        Num: 3, idx: 11
       -Num: 8, idx: 12
       -Num: 8, idx: 13
       -Num: 2, idx: 14
       +Num: 8, idx: 4
       +Num: 8, idx: 4
       +Num: 2, idx: 3
        Num: 0, idx: 15
        Num: 1, idx: 16
       -Num: 3, idx: 17
       -Num: 4, idx: 18
       -Num: 1, idx: 19
       -Num: 1, idx: 20
       -Num: 8, idx: 21
       -Num: 7, idx: 22
       +Num: 3, idx: 11
       +Num: 4, idx: 0
       +Num: 1, idx: 16
       +Num: 1, idx: 16
       +Num: 8, idx: 4
       +Num: 7, idx: 2
        Num: 5, idx: 23
       -Num: 8, idx: 24
       -Num: 5, idx: 25
       -Num: 5, idx: 26
       -Num: 2, idx: 27
       -Num: 0, idx: 28
       -Num: 5, idx: 29
       -Num: 0, idx: 30
       -Num: 2, idx: 31
       -Num: 8, idx: 32
       -Num: 5, idx: 33
       -Num: 6, idx: 34
       -Num: 6, idx: 35
       -Num: 0, idx: 36
       -Num: 5, idx: 37
       -Num: 0, idx: 38
       -Num: 2, idx: 39
       -Num: 0, idx: 40
       -Num: 1, idx: 41
       -Num: 6, idx: 42
       -Num: 6, idx: 43
       -Num: 6, idx: 44
       -Num: 7, idx: 45
       -Num: 4, idx: 46
       -Num: 0, idx: 47
       -Num: 1, idx: 48
       -Num: 8, idx: 49
       -Num: 5, idx: 50
       -Num: 8, idx: 51
       -Num: 8, idx: 52
       -Num: 5, idx: 53
       -Num: 5, idx: 54
       -Num: 5, idx: 55
       -Num: 4, idx: 56
       -Num: 3, idx: 57
       -Num: 7, idx: 58
       -Num: 6, idx: 59
       -Num: 3, idx: 60
       -Num: 2, idx: 61
       -Num: 8, idx: 62
       -Num: 0, idx: 63
       -Num: 2, idx: 64
       -Num: 2, idx: 65
       -Num: 8, idx: 66
       -Num: 1, idx: 67
       -Num: 0, idx: 68
       -Num: 4, idx: 69
       -Num: 0, idx: 70
       -Num: 7, idx: 71
       -Num: 6, idx: 72
       -Num: 1, idx: 73
       -Num: 1, idx: 74
       -Num: 6, idx: 75
       -Num: 2, idx: 76
       -Num: 1, idx: 77
       -Num: 4, idx: 78
       -Num: 8, idx: 79
       -Num: 1, idx: 80
       -Num: 2, idx: 81
       -Num: 8, idx: 82
       -Num: 2, idx: 83
       -Num: 5, idx: 84
       -Num: 3, idx: 85
       -Num: 3, idx: 86
       -Num: 0, idx: 87
       -Num: 6, idx: 88
       -Num: 0, idx: 89
       -Num: 2, idx: 90
       -Num: 6, idx: 91
       -Num: 1, idx: 92
       -Num: 6, idx: 93
       -Num: 1, idx: 94
       -Num: 4, idx: 95
       -Num: 7, idx: 96
       -Num: 4, idx: 97
       -Num: 7, idx: 98
       -Num: 6, idx: 99
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 2, idx: 3
       +Num: 0, idx: 15
       +Num: 5, idx: 23
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 0, idx: 15
       +Num: 5, idx: 23
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 0, idx: 15
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 7, idx: 2
       +Num: 4, idx: 0
       +Num: 0, idx: 15
       +Num: 1, idx: 16
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 8, idx: 4
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 4, idx: 0
       +Num: 3, idx: 11
       +Num: 7, idx: 2
       +Num: 6, idx: 6
       +Num: 3, idx: 11
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 1, idx: 16
       +Num: 0, idx: 15
       +Num: 4, idx: 0
       +Num: 0, idx: 15
       +Num: 7, idx: 2
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 2, idx: 3
       +Num: 1, idx: 16
       +Num: 4, idx: 0
       +Num: 8, idx: 4
       +Num: 1, idx: 16
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 2, idx: 3
       +Num: 5, idx: 23
       +Num: 3, idx: 11
       +Num: 3, idx: 11
       +Num: 0, idx: 15
       +Num: 6, idx: 6
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 4, idx: 0
       +Num: 7, idx: 2
       +Num: 4, idx: 0
       +Num: 7, idx: 2
       +Num: 6, idx: 6
```

There seems to be a problem with the `idx` index being miscalculated. Judging from the `diff` part of the error message, the `to_a.index(i)` (line 19) is returning only the index of the first occurrenc of the element `i`. The index is supposed to be a consecutive value, starting at 0. Your code is trying to find only the index of the first occurrence of the value `i`. Therefor, if there is more than one of any single value in the array, it will return only the index of the first occurrence, instead of a consecutive series of indices. Please fix this.

You may test your code by checking that this produces the desired output:
```
        enum = [1, 1, 1, 2, 2, 2, 3, 3, 3]
        my_each_with_index_output = ''
        block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
        enum.each_with_index(&block)
        each_with_index_output = my_each_with_index_output.dup
        my_each_with_index_output = ''
        enum.my_each_with_index(&block)
        expect(my_each_with_index_output).to eq(each_with_index_output)
```

This should output a series with consecutive indices, and not with indices `[0, 0, 0, 1, 1, 1, 2, 2, 2]`